cinnabar/femap.py

```diff
@@ -0,0 +1,177 @@
+    # absolute Measurements are an edge between 'NULL' and the label
+    # all edges are directed, all edges can be multiply defined
+    computational_graph: nx.MultiDiGraph
+    experimental_graph: nx.MultiDiGraph
+
+    def __init__(self):
+        self.computational_graph = nx.MultiDiGraph()
+        self.experimental_graph = nx.MultiDiGraph()
+
+    @classmethod
+    def from_csv(cls, filename):
+        """Construct from legacy csv format"""
+        data = read_csv(filename)
+
+        # unpack data dictionary
+        fe = cls()
+        for r in data['Calculated']:
+            fe.add_measurement(r)
+        for r in data['Experimental'].values():
+            fe.add_measurement(r)
+
+        return fe
+
+    def add_measurement(self, measurement: Union[RelativeMeasurement, AbsoluteMeasurement]):
+        """Add new observation to FEMap, modifies the FEMap in-place
+
+        Any other attributes on the measurement are used as annotations
+
+        Raises
+        ------
+        ValueError : if bad type given
+        """
+        if isinstance(measurement, AbsoluteMeasurement):
+            # coerce to relative to simplify logic
+            meas_ = RelativeMeasurement(labelA='NULL',
+                                        labelB=measurement.label,
+                                        DDG=measurement.DG,
+                                        uncertainty=measurement.uncertainty,
+                                        computational=measurement.computational)
+        elif isinstance(measurement, RelativeMeasurement):
+            meas_ = measurement
+        else:
+            raise ValueError()
+
+        # slurp out tasty data, anything but labels
+        d = dict(meas_)
+        d.pop('labelA', None)
+        d.pop('labelB', None)
+        d.pop('label', None)
+
+        if meas_.computational:
+            self.computational_graph.add_edge(meas_.labelA, meas_.labelB, **d)
+        else:
+            self.experimental_graph.add_edge(meas_.labelA, meas_.labelB, **d)
+
+    @property
+    def n_measurements(self) -> int:
+        """Total number of both experimental and computational measurements"""
+        return len(self.experimental_graph.edges) + len(self.computational_graph.edges)
+
+    @property
+    def n_ligands(self) -> int:
+        """Total number of unique ligands"""
+        # must ignore NULL sentinel node
+        exptl = self.experimental_graph.nodes - {'NULL'}
+        compt = self.computational_graph.nodes
+
+        return len(exptl | compt)
+
+    @property
+    def degree(self) -> float:
+        """Average degree of all nodes"""
+        return self.n_measurements / self.n_ligands
+
+    def is_weakly_connected(self) -> bool:
+        """Checks if all results in the graph are reachable from other results"""
+        # todo; cache
+        undirected_graph = self.computational_graph.to_undirected()
+        return nx.is_connected(undirected_graph)
+
+    def generate_absolute_values(self):
```

Comment:
to-do post merge: we should try to fix this to work on cases where there are fewer experimental nodes than computational ones, it's a big ask by partners

Reply:
Actually there's nothing to _fix_ here I think. The MLE with will MLE, and should work irrespective of the number of associated experimental nodes. What we probably should methodify is the experimental shift, i.e. https://github.com/OpenFreeEnergy/cinnabar/blob/0f10e11129d5fd2589893794398e528d406b6581/cinnabar/plotting.py#L422-L424

Then folks could call to access shifted dGs if necessary.